src/main/scala/mesosphere/marathon/stream/Subject.scala

```diff
@@ -0,0 +1,117 @@
+package mesosphere.marathon
+package stream
+
+import akka.Done
+import akka.actor.Cancellable
+import akka.stream.OverflowStrategy
+import akka.stream.scaladsl.{Sink => AkkaSink, Source, SourceQueueWithComplete}
+import akka.stream.stage.{GraphStageLogic, InHandler, GraphStageWithMaterializedValue}
+import akka.stream.{Attributes, Inlet, SinkShape}
+import mesosphere.marathon.core.async.ExecutionContexts
+import mesosphere.marathon.util.CancellableOnce
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Internal, stream-indepent logic used for managing the subject. The Subject sink needs to continue to function
+  * after the originating stream terminates. This logic is contained in a vanilla class to prevent usage of any akka
+  * stream methods that would stop working once said stream terminates.
+  */
+private class SubjectInternalLogic[T](queueSize: Int, overflowStrategy: OverflowStrategy) {
+  @volatile private[this] var subscribers: Set[SourceQueueWithComplete[T]] = Set.empty
+  @volatile private[this] var lastElement: T = _
+  @volatile private[this] var elementPushed = false
+  @volatile private[this] var completionResult: Option[Try[Done]] = None
+
+  def onElement(t: T): Unit = synchronized {
+    elementPushed = true
+    lastElement = t
+    subscribers.foreach(_.offer(t))
```

Comment:
Because of our usage of SourceQueue, logic handling downstream stages of the graph will happen in a different thread / outside of this call context. This mitigates the possibility of deadlock here.